### ISSUE TITLE
Possibility to have a loop animation, using triggers. [Feature Request]

### ISSUE THREAD

**Issue by leonardopaiva** [angular/angular#20621]
Hello i`m trying to make a loop animation with full control using triggers, right now theres no easy way to make it, only a few possibles work around.

## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[X] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Not possible to make a loop inside a trigger config, right now i have to check animation done to swap the trigger back and create the loop.

## Expected behavior
Something like a parameter to group, query or animate, setting loop true, or the number of repetitions, animate(300, boolLoopTrue) or animate(300, {boolRepetitions: 3}),  example:
<pre><code>
transition('default => standby', [
      group([
        query('.character', [
           animate(300, style({
            opacity: 0
          })),
          animate(300, style({
            opacity: 1
          }))
        ], boolLoopTrue)
      ])
    ])

</code></pre>

## What is the motivation / use case for changing the behavior?
I`m trying to make complex animations and need it to make things right, stop using work arounds.

## Environment
Angular 
<pre><code>
Angular version: 5.0.0
"@angular/animations": "^5.0.1"

</code></pre>

**Comment by nekkon** [comment]
It would be nice if iterations were defined the following way (where infinite could also be the number of iterations):

````animations: [
   ...
transition('inactive => active', animate('100ms ease-in infinite')),
...
]

**Comment by nekkon** [comment]
@leonardopaiva Have you managed to make an infinite animation? Could you show an example? Unfortunately I tried by looping states but that crashed the browser :(

**Comment by leonardopaiva** [comment]
@nekkon basically you need to use the animation done event that you can attach to the trigger, this event will give you a parameter with the event, it contains the end state, so using that you can swap between start and end states, but this is just a work around. I will up a repository with the example, as soon as possible, i will be back from a trip near January 3 to 5.

Another workaround is to make a basic css animation loop using keyframes, depends a lot on the case.

**Comment by jamesharvey2** [comment]
Exactly. I support this idea. I'm surprised Angular 5 animation feature doesn't ship with built-in loop function. You'll have hard time to properly loop angular animation. CSS3 has loop count built in. You can set a number of times you want your animation to loop. Angular animations should come with the same loop function.

**Comment by jamesharvey2** [comment]
Angular 6 should come with animation loop. Hopefully Google will make this happen.

**Comment by leonardopaiva** [comment]
awesome @jamesharvey2 

@nekkon sry for the delay, but i finally created the loop demo "workaround" you can see it here 
https://angular-animations-demo.firebaseapp.com/loop-demo just click "Go Loop State" the source is here https://github.com/leonardopaiva/angular-animation-basic-project/tree/master/src/app/loop-demo all you need is at loop demo component page.

hope that it helps you or another person with the same need.

**Comment by jamesharvey2** [comment]
amazing... so it uses event fired when animation is finished. 

Still I would like to see 'iteration-count' properly implemented on Angular 6.

**Comment by leonardopaiva** [comment]
yeah @jamesharvey2 this is just a work around, the ideal would be something like:

will make it repeat until the state changes:
```
transition('* => loop-state', [
    style({
      transform: 'translate3d(0,0,0)'
    }),
    animate('300ms', style({
      transform: 'translate3d(300%, 0, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(300%, 100%, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(0, 100%, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(0, 0, 0)'
    }))
  ], {iterationCount: infinite })
```

or 

will make it repeat 3 times 
```
transition('* => loop-state', [
    style({
      transform: 'translate3d(0,0,0)'
    }),
    animate('300ms', style({
      transform: 'translate3d(300%, 0, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(300%, 100%, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(0, 100%, 0)'
    })),
    animate('300ms', style({
      transform: 'translate3d(0, 0, 0)'
    }))
  ], {iterationCount: 3})
```

**Comment by jamesharvey2** [comment]
Yeah, iteration count should be implemented in Angular 6.  I hope Google guys find this feature request.

**Comment by jamesharvey2** [comment]
I would say it's time to abandon angular animations permanently.
Check this new javascript aniamtion framework and see how easily it can trigger complex animations: 

http://animejs.com/documentation/#loop

I realized Angular animations is a waste of time. It's buggy, it has limitations and 'BrowserAnimationsModule' is extremely sensitive. It crashes the whole app with a tiny bit of mistake. 
At this point, I will stop using angular animations until Google team decides to improve it but I doubt if they can create something that can beat Anime.js

**Comment by lazarljubenovic** [comment]
> It crashes the whole app for a tiny bit of mistake.

And thank you for throwing an error instead of auto-magically not working and doing whatever the heck you think you should do to "correct" my mistake.

> I will stop using angular animations until Google team decides to improve it

You do realize that Angular is open-source? You're free to jump in to help "create something that can beat Anime.js".

**Comment by MrCroft** [comment]
@jamesharvey2 
Let's say we would drop Angular Animations and design all animations with animejs instead... to get the loop functionality one desires, without having to write a workaround until it lands in a future 6.x version.
But then we'd have to write more code to:
1. Take care of SSR (because animejs works directly with the DOM) - not sure if Domino would flawlessly handle animejs;
2. Properly handle routing animations;
3. Create stagger animations;
4. Handle `:enter`, `:leave` and such;
5. Probably there are more scenarios, these are just off the top of my head... The animations api in Angular is much more powerfull than animejs and, also, it's designed to work **with** Angular.

By the way, I see that the [Animation Player](https://angular.io/api/animations/AnimationPlayer) has a **restart** method. Haven't tried it, but If it does what it intuitively says it does, you could use the player to handle a looping animation and `restart()` on `onDone` (if a condition is met, of course). Maybe that would be less code than other workarounds already in use out there.

OR I guess... ultimately you could use plain old css to create that one infinite loop animation you need in a project (not an `n` loop, for that you should probably use the Animation Player) rather than bringin in an entire library and then have headaches working to implement the above mentioned points without Angular Animations.

**Comment by Swoox** [comment]
I'm not sure if it will help somebody out but I have found a really nice infinity solution.
I had seen this nice rotation [rotate ball](https://stackblitz.com/edit/angular-image-rotation?file=app%2Fapp.component.html) . What I did was add the following to the `<img`: `(@rotatedState.done)="rotate()`. This will make the animation infinity.

**Comment by DmitryEfimenko** [comment]
Here's how I implemented repeating an animation n times:
```ts
@ViewChild('animateEl', { read: ElementRef }) animateEl: ElementRef;

constructor(private _builder: AnimationBuilder) { }

private runAnimationNTimes(n: number) {
  // IMPORTANT: for some reason `const animation` needs to be declared in this function
  // If it's declared outside, animation does not work properly.
  const animation = animate('10s ease-out', keyframes([
    style({ offset: 0, transform: 'rotate(0)' }),
    // more styles here...
  ]));

  const factory = this._builder.build(animation);
  const player = factory.create(this.animateEl.nativeElement);
  player.onDone(() => {
    player.destroy();
    if (n > 0) {
      this.runAnimationNTimes(--n);
    }
  });

  player.play();
}
```

**Comment by xxzhong21330** [comment]
it's so weird. when i use AnimationBuilder Service to create animationPlayer, the AnimationPlayer's beforeDestroy callback is never invoked with executing the animationPlayer. destroy() synchronously, but i execute animationPlayer.destroy() asynchronously, everything looks normal, here's how implemented infinite loop:
```
@ViewChild('el', { read: ElementRef }) el: ElementRef;
let myAnimation = this._builder.build(
    animate( '2s', keyframes([
        style({transform: `translate3d(0,0,0)`, offset: 0}),
        style({transform: `translate3d(-100px,0,0)`, offset: 1}),
    ]))
);
let animationPlay = ()=>{
    let player = myAnimation.create(this.el.nativeElement);
    player.play();
    player.onDone(()=>{
        setTimeout(()=>{
            player.destroy();
        })
    });
    player.onDestroy(()=>{
        animationPlay();
    });
}
animationPlay();
```

**Comment by AliSawari** [comment]
@leonardopaiva any updates on the requested feature ? this should have been  implemented by now :neutral_face:

**Comment by nektobit** [comment]
Hi
Any news?

**Comment by bobjane1234** [comment]
Hello any news?

**Comment by RiotPharaoh** [comment]
Hello, any news?

**Comment by lazarljubenovic** [comment]
Yes, there are news, but the team is purposely not giving them away because they decided the best course action was never to release the code to the public and not to tell us about animation loops.

**Comment by LaXiS96** [comment]
Angular 15 breaks CSS `@keyframes`, which are currently the only clean and obvious way to get infinite looping animations.
None of the proposed migrations are without drawbacks (putting keyframes in the global style or changing view encapsulation) and the suggested one (using Angular animations) does not support infinite loops.
I can't update my application to Angular 15.

**Comment by JDavid21051** [comment]
Hi
Any news?

**Comment by VladimirCreator** [comment]
7 years)

**Comment by Senneseph** [comment]
any questions, just ask.  before god there was nothing and from nothing you can ger two nothings.
art attack attack on the heart, the shape of the hole in my heart. ideas well, i don't have an exact question answer for that, but. those two nothings were a point and a halo and from that god created everything
Ask whoever authored moody blues, flaming lips, there is a story of humanity, one of inspiration almost everything from lored of the rings to see as lewis aka le wis.

**Comment by Senneseph** [comment]
sorry, not halo, water droplet

**Comment by Senneseph** [comment]
sorry, still unwinding.
before the water droplet was nothing and so tye bit of light was lonely

**Comment by Senneseph** [comment]
so vs to verse how tell them apart? japanese has incredible phonetic groupings but...
there's more to tell.

sorry, the light could only see itself in a water froplet and iy thought it was itself

**Comment by Senneseph** [comment]
sorry, thought it was an other (brother)
well get to madame im atom
(adam made me)

**Comment by Senneseph** [comment]
there are so many syories but the squirming tail ofnsumset is quite tricky

**Comment by Senneseph** [comment]
apologies for the typ os, its the only way to type os

**Comment by Senneseph** [comment]
type Os

**Comment by Senneseph** [comment]
case matters

**Comment by Senneseph** [comment]
Anyway, this story goes on like this. The spec of flight c its reflection and I thought it was another, but no matter how fast it race, it just couldn't be the other. Yeah, see. Aha. They all work, don't worry. I'll make mistakes because II don't have any rehearsal. I just got excited, but He's hot, Helium

**Comment by Senneseph** [comment]
got that fibre always a round m  hi im these go pn phorever and ever

**Comment by Senneseph** [comment]
don't go on?

**Comment by Senneseph** [comment]
that why the upside down ? needes a mirror
to separtate? from two say pay rate

**Comment by Senneseph** [comment]
part8

**Comment by Senneseph** [comment]
this keyboard is terrible to me

**Comment by Senneseph** [comment]
Ha ha.If you follow the yellow rick road, you'll find it in every story ever written.I'm pickle, rick

**Comment by Senneseph** [comment]
there is a hxidden path
hidden meaning blocked
change aft most to most i'm her

**Comment by Senneseph** [comment]
seraphim are the tradematks of got they are the plutonim plutonium platonic solids lids so not ic3yic to play 
look you jusy removed yhe ser ifs an other test

**Comment by Senneseph** [comment]
klop, plunk dunk stunk trunk monk uncool you no see all everything forever

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
